File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> combination;
        combinationSum(candidates, target, res, combination, 0);
        return res;
    }

private:
    void combinationSum(vector<int>& candidates, int target, vector<vector<int>>& res, vector<int>& combination, int begin) {
        if (!target) {
            res.push_back(combination);
            return;
        }
        for (int i = begin; i != candidates.size() && target >= candidates[i]; ++i) {
            if (i == begin || candidates[i] != candidates[i - 1]) {
                combination.push_back(candidates[i]);
                combinationSum(candidates, target - candidates[i], res, combination, i + 1);
                combination.pop_back();
            }
        }
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.cs
public class Solution {
    public IList<IList<int>> CombinationSum2(int[] candidates, int target) {
        Array.Sort(candidates);
        var result = new List<IList<int>>();
        CombinationSum2(candidates, target, 0, new List<int>(), result);
        return result;
    }

    private void CombinationSum2(int[] candidates, int target, int start, List<int> combination, IList<IList<int>> result) {
        if (target == 0) {
            result.Add(new List<int>(combination));
            return;
        }

        for (int i = start; i < candidates.Length && candidates[i] <= target; i++) {
            if (i > start && candidates[i] == candidates[i - 1]) continue; // skip duplicates

            combination.Add(candidates[i]);
            CombinationSum2(candidates, target - candidates[i], i + 1, combination, result);
            combination.RemoveAt(combination.Count - 1); // remove last element
        }
    }
}
